src/java/org/apache/cassandra/index/sai/disk/vector/VectorMemtableIndex.java

```diff
@@ -164,64 +160,56 @@ else if (primaryKey.compareTo(maximumKey) > 0)
-            // if right token is MAX (Long.MIN_VALUE), there is no upper bound
-            boolean isMaxToken = keyRange.right.getToken().isMinimum(); // max token
-
-            PrimaryKey left = indexContext.keyFactory().createTokenOnly(keyRange.left.getToken()); // lower bound
-            PrimaryKey right = isMaxToken ? null : indexContext.keyFactory().createTokenOnly(keyRange.right.getToken()); // upper bound
+        float threshold = expr.getEuclideanSearchThreshold();
 
-            Set<PrimaryKey> resultKeys = isMaxToken ? primaryKeys.tailSet(left, leftInclusive) : primaryKeys.subSet(left, leftInclusive, right, rightInclusive);
+        var pkIterator = searchInternal(context, qv, keyRange, graph.size(), threshold,
+                                        (keys) -> this.filterPrimaryKeys(qv, threshold, keys),
+                                        this::nodeScoreToPrimaryKey);
 
-            if (resultKeys.isEmpty())
-                return RangeIterator.empty();
-
-            int bruteForceRows = maxBruteForceRows(limit, resultKeys.size(), graph.size());
-            logger.trace("Search range covers {} rows; max brute force rows is {} for memtable index with {} nodes, LIMIT {}",
-                         resultKeys.size(), bruteForceRows, graph.size(), limit);
-            Tracing.trace("Search range covers {} rows; max brute force rows is {} for memtable index with {} nodes, LIMIT {}",
-                          resultKeys.size(), bruteForceRows, graph.size(), limit);
-            if (resultKeys.size() <= bruteForceRows)
-                return new ReorderingRangeIterator(new PriorityQueue<>(resultKeys));
-            else
-                bits = new KeyRangeFilteringBits(keyRange);
-        }
+        PriorityQueue<PrimaryKey> keyQueue = new PriorityQueue<>();
+        while (pkIterator.hasNext())
+            keyQueue.add(pkIterator.next());
 
-        var keyQueue = graph.search(qv, limit, expr.getEuclideanSearchThreshold(), bits);
         if (keyQueue.isEmpty())
             return RangeIterator.empty();
         return new ReorderingRangeIterator(keyQueue);
     }
 
     @Override
-    public OrderIterator orderBy(QueryContext queryContext, Expression expr, AbstractBounds<PartitionPosition> keyRange, int limit)
+    public ScoredPrimaryKeyIterator orderBy(QueryContext context, Expression expr, AbstractBounds<PartitionPosition> keyRange, int limit)
     {
         assert expr.getOp() == Expression.Op.ANN : "Only ANN is supported for vector search, received " + expr.getOp();
 
+        float[] qv = expr.lower.value.vector;
+
+        var result = searchInternal(context, qv, keyRange, limit, 0,
+                                    (keys) -> orderPrimaryKeys(qv, keys),
+                                    this::nodeScoreToScoredPrimaryKey);
+        return new WrappedScoredPrimaryKeyIterator(result);
+    }
+
+    private Iterator<PrimaryKey> nodeScoreToPrimaryKey(SearchResult.NodeScore nodeScore)
```

Comment:
It looks like we could remove a bunch of code if we just standardized on using ScoredPrimaryKey instead of bare PK, even when we don't "need" the score, and our history has been towards finding more places that score will be useful rather than less

Reply:
That's true. A fair amount of the extra code in this PR (especially the memtable index code) is present because I was being careful about object creation. If we don't mind wrapping all PKs in a `ScoredPrimaryKey`, then we can collapse it now.

Reply:
I think this is a case of YAGNI, we can optimize later if the profiler tells us to

Reply:
Works for me. The pulsar community was very concerned about object creation and unnecessary boxing, so you're seeing that here.